[WIP] RHDEVDOCS-3917 - documents new authentication methods for remote_write in monitoring

bburt-rh opened the pull request:
[WIP]
Summary: This PR documents additional endpoint authentication methods for the remote_write feature in OCP monitoriing.

- Aligned team: DevTools
- For branches: 4.11
- Jira: https://issues.redhat.com/browse/RHDEVDOCS-3917
- Direct link to doc preview: tbd
- SME review: @ tbd
- QE review: @ tbd
- Peer review: @ tbd

bburt-rh commented:
@raptorsun @lihongan @juzhao @JoaoBraveCoding Please review this WIP PR for configuring additional auth methods for remote_write

lihongan commented:
@lihongyan1  ^^

lihongyan1 commented:
thx

On Wed, Apr 27, 2022 at 9:39 AM Hongan Li ***@***.***> wrote:

> @lihongyan1 <https://github.com/lihongyan1> ^^
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/openshift-docs/pull/44713#issuecomment-1110438697>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOLLYMNRJUXVYQJJOJFURULVHCLEXANCNFSM5TZ4I2YA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -53,25 +54,52 @@ data:
 +
-* The following example configures basic authentication:
+The following shows a `sigv4` sample configuration.
+It assumes you have already created a `sigv4` secret named `sigv4-credentials`in the `openshift-monitoring` namespace.
```
`sigv4-credentials`in 
should be
`sigv4-credentials` in

bburt-rh replied:
fixed

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -53,25 +54,52 @@ data:
+data:
+  config.yaml: |
+    prometheusK8s:
+      remoteWrite:
+       - url: "https://authorization.remotewrite.com/api/write"
```
```
      remoteWrite:
       - url: "https://authorization.remotewrite.com/api/write"
```
it is better change to
```
      remoteWrite:
      - url: "https://authorization.remotewrite.com/api/write"
```

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -53,25 +54,52 @@ data:
+Basic authorization (`basicAuth`) sets the authorization header on every remote write request with the configured username and password.
++
+The following sample shows a `basicAuth` configuration.
+It assumes you have already set up authentication credentials for the endpoint.
```
change to
It assumes you have already set up authentication credentials named `remoteWriteAuth` for the endpoint.

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -88,27 +116,93 @@ data:
         basicAuth:
           username:
             name: remoteWriteAuth
```
we should mention what's remoteWriteAuth here.
FYI, for secret, the secret name is improper, since
```
The Secret "remoteWriteAuth" is invalid: metadata.name: Invalid value: "remoteWriteAuth": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

bburt-rh replied:
- changed name from `remoteWriteAuth` to `rw-basic-auth`
- Added a sentence saying `It also assumes that you have already created a `basicAuth` secret with the name `rw-basic-auth` in the `openshift-monitoring` namespace.`

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -88,27 +116,93 @@ data:
+* OAuth 2.0
++
+An OAuth 2.0 (`oauth2`) configuration uses the client credentials grant type. Prometheus fetches an access token from the specified remote write endpoint with the specified client access and secret keys.
+You cannot use this method simultaneously with `basic_auth`, `Bearer`, or `sigv2` authentication.
```
basic_auth -> basicAuth
sigv2 -> sigv4

bburt-rh replied:
fixed

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -88,27 +116,93 @@ data:
+              - <scope_2>
+            endpointParams:
+              param1: <value_1>
+              param2: <value_2>
```
it's better to change the format to below, there are additional spaces
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: https://test.remotewrite.com/api/write
        remoteTimeout: 30s
        oauth2:
          clientId:
            secret:
              name: oauth2-credentials
              key: <id>
          clientSecret:
            name: oauth2-credentials
            key: <secret>
          tokenUrl: https://example.com/oauth2/token
          scopes:
          - <scope_1>
          - <scope_2>
          endpointParams:
            param1: <value_1>
            param2: <value_2>
```

bburt-rh replied:
fixed

juzhao replied:
see format is still wrong, please see my example

JoaoBraveCoding commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -88,27 +116,93 @@ data:
+  config.yaml: |
+    prometheusK8s:
+      remoteWrite:
+      - url: "https://remote-write.endpoint"
```
Not sure, if in this example, we care about the first two entries of this array since only the last entry contains information actually relevant for user wanting to use this method of configuration

bburt-rh replied:
deleted lines 144 to 152 and changed the descriptive text

raptorsun commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -53,25 +54,52 @@ data:
-For `endpoint_authentication_credentials` substitute the credentials for the endpoint.
-Currently supported authentication methods are basic authentication (`basicAuth`) and client TLS (`tlsConfig`) authentication.
+<1> For `endpoint_authentication_credentials` substitute the authentication method and the credentials for the endpoint.
+Currently supported authentication methods are AWS Signature version 4 (`sigv4`), basic authentication (`basicAuth`),  bearer token (`Bearer`), TLS client (`tlsConfig`), and OAuth 2.0 (`oauth2`). 
```
There is the field `bearerTokenFile` to specify the bearer token. The file should be mounted to the Prometheus pods.
https://github.com/openshift/cluster-monitoring-operator/blob/release-4.11/pkg/manifests/config.go#L158

bburt-rh replied:
Changed all occurrences of `Bearer` to `bearerTokenFile`

juzhao replied:
lgtm

juzhao replied:
the doc is for epic https://issues.redhat.com/browse/MON-2160, there is one task
https://issues.redhat.com/browse/MON-2207 Expose Authorization settings for remote write in the CMO configuration
`authorization` is defined in
https://github.com/openshift/cluster-monitoring-operator/blob/release-4.11/pkg/manifests/config.go#L160
`Authorization *monv1.SafeAuthorization `json:"authorization,omitempty"``

check SafeAuthorization struct, Type default value is Bearer, so I think `Bear` is fine, but `bearerTokenFile` is also fine
https://github.com/openshift/cluster-monitoring-operator/blob/release-4.11/vendor/github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1/types.go#L1752-L1758
```
type SafeAuthorization struct {
	// Set the authentication type. Defaults to Bearer, Basic will cause an
	// error
	Type string `json:"type,omitempty"`
	// The secret's key that contains the credentials of the request
	Credentials *v1.SecretKeySelector `json:"credentials,omitempty"`
}
```

JoaoBraveCoding replied:
> I changed all occurrences of Bearer to bearerTokenFile. 

@bburt-rh this is not correct. In the RemoteWrite we have essentially 3 ways of doing authentication:
1. BasicAuth to use this method users have the field `remoteWrite.basicAuth`;
2. Bearer token to use this method users can provide either a `bearerTokenFile` with the field `remoteWrite.bearerTokenFile`  alternatively or specify a secret where the bearer token is located with `remoteWrite.authorization` where `Type: Bearer`.
3. Sigv4 to use this method users have the field `remoteWrite.sigv4`

Hopefully this made it more clear. If you want examples in this link you have test cases for all the methods [1]
[1] https://github.com/openshift/cluster-monitoring-operator/blob/dc7b00300f9a8c0d9cfde08724b9ef58292bab48/pkg/manifests/manifests_test.go#L1271

bburt-rh replied:
I've documented the `Bearer` and `bearerFile` methods of configuring bearer token authorization.

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -88,27 +116,93 @@ data:
+            name: rw-auth
+            key: <password>
+      - url: "https://authorization.remotewrite.com/api/write"
+        authorization:
```
from https://prometheus.io/docs/prometheus/latest/configuration/configuration
I see
```
# Authentication information used to authenticate to the API server.
# Note that `basic_auth`, `authorization` and `oauth2` options are
# mutually exclusive.
```
then we should not configure `basicAuth` and `authorization` together

juzhao replied:
not sure about it, needs confirmation from developer

bburt-rh replied:
deleted lines 144 to 152 and changed the descriptive text

lihongyan1 commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -88,27 +116,93 @@ data:
+A TLS client (`tlsConfig`) configuration specifies the CA certificate, the client certificate, and the client key file information used to authenticate with the remote write endpoint server using TLS.
++
+The following sample shows a `tlsConfig` configuration.
+It assumes that you have already created a CA certificate file, a client certificate file, and a client key file.
```
It assumes you have already created a CA certificate file, a client certificate file, and a client key file. You also have created a `tls` secret with the name `selfsigned-mtls-bundle` in the `openshift-monitoring` namespace.

bburt-rh replied:
Added the new sentence

lihongyan1 commented:
All the authentication methods need assume a correspond secret have created, it'd be better give the yaml or command in the doc about how to create the secret.

raptorsun commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -88,27 +116,93 @@ data:
 ----
+* OAuth 2.0
++
+An OAuth 2.0 (`oauth2`) configuration uses the client credentials grant type. Prometheus fetches an access token from the specified remote write endpoint with the specified client access and secret keys.
```
Prometheus fetches an access token from `tokenUrl`  with the specified client name and client secret, in order to access the remote write endpoint using this token.

bburt-rh replied:
Updated the content per your comment

juzhao commented:
> All the authentication methods need assume a correspond secret have created, it'd be better give the yaml or command in the doc about how to create the secret.

I don't think we should explain too explicit

lihongyan1 commented:
> > All the authentication methods need assume a correspond secret have created, it'd be better give the yaml or command in the doc about how to create the secret.
> 
> I don't think we should explain too explicit

When we need remote write metric data to a target, suppose we need complete the following:
1. Figure out remote target information, include api url and authentication
2. Create secret based on the information
3. Edit configmap to put in the api url and secret name, secret keys created in the step 2
So I think we'd better provide information about how to create secret.

bburt-rh commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -53,25 +54,52 @@ data:
+* AWS Signature Version 4
++
+This method uses AWS Signature Version 4 (`sigv4`) authentication to sign requests.
+This method cannot be used simultaneously with `basic_auth`, `Bearer`, or `oauth2`.
```
fixed

bburt-rh commented:
@lin

> > > All the authentication methods need assume a correspond secret have created, it'd be better give the yaml or command in the doc about how to create the secret.
> > 
> > 
> > I don't think we should explain too explicit
> 
> When we need remote write metric data to a target, suppose we need complete the following:
> 
> 1. Figure out remote target information, include api url and authentication
> 2. Create secret based on the information
> 3. Edit configmap to put in the api url and secret name, secret keys created in the step 2
>    So I think we'd better provide information about how to create secret.

@lihongyan1 and @juzhao We list the following requirement in the prerequisites:

`* You have set up authorization secrets for the remote target that include the API URL and the authentication method.`

Is this enough information to include?

lihongyan1 commented:
> > > > All the authentication methods need assume a correspond secret have created, it'd be better give the yaml or command in the doc about how to create the secret.
> > > 
> > > 
> > > I don't think we should explain too explicit
> > 
> > 
> > When we need remote write metric data to a target, suppose we need complete the following:
> > 
> > 1. Figure out remote target information, include api url and authentication
> > 2. Create secret based on the information
> > 3. Edit configmap to put in the api url and secret name, secret keys created in the step 2
> >    So I think we'd better provide information about how to create secret.
> 
> @lihongyan1 and @juzhao We list the following requirement in the prerequisites:
> 
> `* You have set up authorization secrets for the remote target that include the API URL and the authentication method.`
> 
> Is this enough information to include?

For these different authentication ways, need different secret types(opaque or tls) and the configmap need use previously created secret name and keys, it would be more clear and easy to follow if provide detail of secret.

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -126,14 +212,14 @@ data:
           keySecret:
             name: selfsigned-mtls-bundle
-            key: client.key
+            key: <client.key>
```
see from https://deploy-preview-44713--osdocs.netlify.app/openshift-enterprise/latest/monitoring/configuring-the-monitoring-stack.html#configuring_remote_write_storage_configuring-the-monitoring-stack
the `tlsConfig` configmap format is different with others

bburt-rh replied:
fixed

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -87,29 +116,86 @@ data:
+        authorization:
+          type: bearerTokenFile
+          credentials:
+            name: rw-berer-auth
```
rw-bearer-auth

bburt-rh replied:
fixed

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -87,29 +116,86 @@ data:
-* The following example configures client TLS authentication:
+* *bearer token* 
++
+A bearer token (`bearerTokenFile`) configuration combines a basic authentication username and password with authorization credentials configured to use a specified token (such as an API key).
```
I think the description is improper, since bearer token sets the `Authorization` header on every scrape request with the configured credentials.

JoaoBraveCoding commented on modules/monitoring-example-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,340 @@
+== Sample for bearer token authentication
+
+Bearer token authentication sets the authorization header on every remote write request using the configured token.
+This method can be used together with basic authentication.
```
Although this might be true, I don't see the point of adding the basicAuth configuration to the bearer token example. Do we have a reason to do it this way?

bburt-rh replied:
deleted the basicAuth section from this sample

JoaoBraveCoding commented:
Small comment, but I think that these lines [1] should probably be moved somewhere else, since it doesn't make much sense to me that when telling users how they can modify their remote write config to in the middle talk about UWM config. I would say to either completely remove it since we have a mention of it the prerequisites or move it to the beggining/end of this procedure. WDYT?

[1] https://github.com/openshift/openshift-docs/pull/44713/files#diff-3aab0fa7f4c559320905597f8089d88d5044f6c23e132a589badfa795e006cceL182-L203

JoaoBraveCoding commented on monitoring/configuring-the-monitoring-stack.adoc:
```diff
@@ -108,13 +108,16 @@ include::modules/monitoring-modifying-retention-time-for-prometheus-metrics-data
 * See link:https://prometheus.io/docs/practices/remote_write/#remote-write-tuning[Tuning remote write settings] for information about how to optimize remote write settings for different use cases.
-* For information about additional optional fields, please refer to the API documentation.
+* See xref:../nodes/pods/nodes-pods-secrets.adoc#nodes-pods-secrets-about_nodes-pods-secrets[Understanding secrets] for steps to create and configure `Secret` objects in {product-title}.
+* For information about additional optional fields, see the xref:../rest_api/monitoring_apis/prometheus-monitoring-coreos-com-v1.adoc[Prometheus REST API reference].
```
Here, WDYT about linking instead the API Reference only for the Spec.RemoteWrite part? [1]

[1] https://deploy-preview-44713--osdocs.netlify.app/openshift-enterprise/latest/rest_api/monitoring_apis/prometheus-monitoring-coreos-com-v1.html#spec-remotewrite-2

bburt-rh replied:
Yes. That makes sense. I'll change it.

bburt-rh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,330 @@
+----
+<1> The authentication token.
+
+The following shows sample bearer token config map settings that use a `Secret` object named `rw-bearer-auth`in the `openshift-monitoring` namespace:
```
fixed

bburt-rh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,330 @@
+<4> The unique identifier for the Amazon Resource Name (ARN) assigned to your role.
+
+
+.Sample for basic authentication
```
fixed

bburt-rh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,330 @@
+----
+<1> Path to the bearer token file.
+
+.Sample for bearer token authentication using a `Secret` Object
```
fixed

bburt-rh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,330 @@
+It assumes that you have already set up authentication credentials for the endpoint. 
+
+[source,yaml]
+----
```
good point. Fixed

bburt-rh commented:
> Small comment, but I think that these lines [1] should probably be moved somewhere else, since it doesn't make much sense to me that when telling users how they can modify their remote write config to in the middle talk about UWM config. I would say to either completely remove it since we have a mention of it the prerequisites or move it to the beggining/end of this procedure. WDYT?
> 
> [1] https://github.com/openshift/openshift-docs/pull/44713/files#diff-3aab0fa7f4c559320905597f8089d88d5044f6c23e132a589badfa795e006cceL182-L203

@JoaoBraveCoding  The link just takes me to the "Files changed" page. Can you put a comment in the code for the content that you are referring to or point me to a line number in a file? Is it the NOTE admonition about UWM on the samples page? Thanks!

bburt-rh commented:
> For these different authentication ways, need different secret types(opaque or tls) and the configmap need use previously created secret name and keys, it would be more clear and easy to follow if provide detail of secret.

@lihongyan1 @juzhao @raptorsun @JoaoBraveCoding I've revised and reorganized the content in this PR substantially:

- I've added sample YAML for the corresponding `Secret` object required for each auth method. 
- I've added a table describing all supported auth methods for remote write.
- I've made the remote write sample YAML more consistent.

It's now ready for another review. Thanks!

JoaoBraveCoding commented:
> @JoaoBraveCoding The link just takes me to the "Files changed" page. Can you put a comment in the code for the content that you are referring to or point me to a line number in a file? Is it the NOTE admonition about UWM on the samples page? Thanks!

Sorry @bburt-rh still learning GitHub specifics... I, somehow, am not able to link the part I want, so here is a screen shot [1] it's in file monitoring-configuring-remote-write-storage.adoc starting at line 182.

[1] 
![image](https://user-images.githubusercontent.com/10974558/166467243-554cc482-ff32-4bae-a9b2-e84fc31a13da.png)

raptorsun commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,331 @@
+      - url: "https://test.example.com/api/write"
+        oauth2:
+          clientId:
+            secret:
```
shall we mention the fields such as `clientId` can either be a secret or configmap?
same for the fields in tls config: `ca`, `cert`. 
`clientSecret` and `keySecret` must be secret though.

bburt-rh commented:
> Sorry @bburt-rh still learning GitHub specifics... I, somehow, am not able to link the part I want, so here is a screen shot [1] it's in file monitoring-configuring-remote-write-storage.adoc starting at line 182.
> 
> [1] ![image](https://user-images.githubusercontent.com/10974558/166467243-554cc482-ff32-4bae-a9b2-e84fc31a13da.png)

No worries, @JoaoBraveCoding. I've updated the content and moved the UWM info to a NOTE at the beginning of the procedure. PTAL and let me know whether or not this addresses your concern. Thanks!

JoaoBraveCoding commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,331 @@
+----
+<1> The URL of the remote write endpoint.
+<2> The required configuration details for the authentication method for the endpoint.
+Currently supported authentication methods are Amazon Web Services (AWS) Signature Version 4, bearer token, basic authentication, OAuth 2.0, and TLS client. 
```
Maybe here it also makes sense to add the note about the UWM ConfigMap just for good measure, WDYT?

bburt-rh replied:
Thanks @JoaoBraveCoding. That's a good idea. I've added the following admonition here:

```
[NOTE]
====
If you want to configure remote write for the Prometheus instance that monitors user-defined projects, edit the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace. Note that TLS client configuration for remote write is not supported for use with user-defined projects.
====
```

JoaoBraveCoding reviewed (approved):
Apart from the small comment, LGTM good job @bburt-rh

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -52,89 +60,14 @@ data:
+<1> The URL of the remote write endpoint.
+<2> The authentication method and credentials for the endpoint.
+Currently supported authentication methods are AWS Signature Version 4, bearer token, basic authentication, OAuth 2.0, and TLS client. 
+See _Supported remote write authentication settings_ below for sample configurations of supported authentication methods.
```
"Supported remote write authentication settings" should add a link to the help part

bburt-rh replied:
@juzhao OpenShift documentation standards do not allow me to add a cross-reference link from this module file to another module file. See the "Modules" section of this doc fyi: https://docs.google.com/document/d/1WN7k72PKxcPA__erp5TK-CefmS_sVwE3r40PvpmNk3E/edit#heading=h.k4v49gmkfxi

juzhao replied:
ok, no question

juzhao commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -52,89 +60,14 @@ data:
+Currently supported authentication methods are AWS Signature Version 4, bearer token, basic authentication, OAuth 2.0, and TLS client. 
+See _Supported remote write authentication settings_ below for sample configurations of supported authentication methods.
 
 . Add write relabel configuration values after the authentication credentials:
```
I think this setting is optional for configure remote write storage, but keep it is fine

raptorsun reviewed (approved):
/lgtm

lihongyan1 commented:
/lgtm

lihongyan1 commented:
> > For these different authentication ways, need different secret types(opaque or tls) and the configmap need use previously created secret name and keys, it would be more clear and easy to follow if provide detail of secret.
> 
> @lihongyan1 @juzhao @raptorsun I've revised and reorganized the content in this PR substantially:
> 
> * I've added sample YAML for the corresponding `Secret` object required for each auth method.
> * I've added a table describing all supported auth methods for remote write.
> * I've made the remote write sample YAML more consistent.
> 
> It's now ready for another review. Thanks!

Great! thanks!

juzhao commented:
/lgtm

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,330 @@
+----
+<1> The name of the `Secret` object that contains the authentication credentials.
+
+.Sample YAML for bearer token authentication using a token file
```
as I commented above, I believe that we only need an example for `bearer`.

bburt-rh replied:
deleted

simonpasquier reviewed (commented):
Wow this is very exhaustive!

bburt-rh commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -21,15 +21,23 @@ Doing so has no impact on how or for how long Prometheus stores metrics.
 * You have set up a remote write compatible endpoint (such as Thanos) and know the endpoint URL.
 See the link:https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage[Prometheus remote endpoints and storage documentation] for information about endpoints that are compatible with the remote write feature.
-* You have set up authentication credentials for the remote write endpoint. 
+* You have set up authentication credentials in a `Secret` object for the remote write endpoint.
```
Added this content to clarify the namespace for the secret needs to be the same as the Prometheus object

bburt-rh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,330 @@
+            param2: <value_2>
+----
+<1> The name of the corresponding `Secret` object. Note that `ClientId` can alternatively refer to a `ConfigMap` object, though `clientSecret` must refer to a `Secret` object.
+<2> URL used to access a token with the specified `clientId` and `clientSecret`.
```
changed `access` to `fetch`

bburt-rh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,330 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: selfsigned-mtls-bundle
```
deleted `selfsigned-` throughout

bburt-rh commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -21,15 +21,23 @@ Doing so has no impact on how or for how long Prometheus stores metrics.
+
+[NOTE]
+====
+If you want to configure remote write for the Prometheus instance that monitors user-defined projects, make similar edits to the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace
```
Added this content back and put it at the end of the note admonition.

bburt-rh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,330 @@
+          region: <AWS_region> <1>
+          accessKey:
+            name: sigv4-credentials <2>
+            key: access
```
Good idea. I added in callout footnotes for all of the key fields in all of the code samples

rolfedh commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -21,15 +21,25 @@ Doing so has no impact on how or for how long Prometheus stores metrics.
 See the link:https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage[Prometheus remote endpoints and storage documentation] for information about endpoints that are compatible with the remote write feature.
-* You have set up authentication credentials for the remote write endpoint. 
+* You have set up authentication credentials in a `Secret` object for the remote write endpoint.
+The secret needs to be created in the same namespace as the Prometheus object for which you configure remote write: `openshift-monitoring` for default platform monitoring or `openshift-user-workload-monitoring` for user workload monitoring.
```
```suggestion
["You must create the secret" OR "The secret must be"] in the same namespace as the Prometheus object for which you configure remote write: `openshift-monitoring` for default platform monitoring or `openshift-user-workload-monitoring` for user workload monitoring.
```

rolfedh replied:
I wonder if we should unpack the meaning a bit to make it easier for non-fluent readers. Perhaps:
```suggestion
The secret must be in the same namespace as the Prometheus object for which you configure remote write. That is, for default platform monitoring, the secret must be in the `openshift-monitoring` namespace. Or, for user workload monitoring, the secret must be in the `openshift-user-workload-monitoring` namespace.
```

rolfedh commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -21,15 +21,25 @@ Doing so has no impact on how or for how long Prometheus stores metrics.
+[NOTE]
+====
+If you want to configure remote write for the Prometheus instance that monitors user-defined projects, make similar edits to the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace. 
+If you do so, note that the Prometheus config map component is called `prometheusK8s` in the `cluster-monitoring-config` `ConfigMap` object and `prometheus` in the `user-workload-monitoring-config` `ConfigMap` object.
```
```suggestion
If you do so, note that the Prometheus config map component is called `prometheusK8s` in the `cluster-monitoring-config` config map and `prometheus` in the `user-workload-monitoring-config` config map.
```

bburt-rh replied:
Revised to read `Note that the Prometheus config map component is called `prometheus` in the `user-workload-monitoring-config` `ConfigMap` object and not `prometheusK8s`, as it is in the `cluster-monitoring-config` `ConfigMap` object.`

rolfedh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+            key: access <3>
+          secretKey:
+            name: sigv4-credentials <2>
+            key: secret <3>
```
Should we represent any of the values in the preceding five lines be between angle brackets?

bburt-rh replied:
No. These are the actual values. They correspond to the fields in the `Secret` object that contain the credentials.

rolfedh commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -21,15 +21,25 @@ Doing so has no impact on how or for how long Prometheus stores metrics.
 [CAUTION]
 ====
-To reduce security risks, avoid sending metrics to an endpoint via unencrypted HTTP or without using authentication.
+To reduce security risks, avoid sending metrics to an endpoint using unencrypted HTTP or without using authentication.
```
Maybe
```suggestion
To reduce security risks, use HTTPS and authentication to send metrics to an endpoint.
```

bburt-rh replied:
revised per your suggestion

rolfedh commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -21,15 +21,25 @@ Doing so has no impact on how or for how long Prometheus stores metrics.
+
+[NOTE]
+====
+If you want to configure remote write for the Prometheus instance that monitors user-defined projects, make similar edits to the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace. 
```
```suggestion
If you configure remote write for the Prometheus instance that monitors user-defined projects, make similar edits to the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace. 
```

bburt-rh replied:
revised per your suggestion

rolfedh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+
+[NOTE]
+====
+If you want to configure remote write for the Prometheus instance that monitors user-defined projects, edit the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace.
```
```suggestion
If you configure remote write for the Prometheus instance that monitors user-defined projects, edit the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace.
```

bburt-rh replied:
Fixed and added a sentence: `Note that the Prometheus config map component is called `prometheus` in the `user-workload-monitoring-config` `ConfigMap` object and not `prometheusK8s`, as it is in the `cluster-monitoring-config` `ConfigMap` object. `

rolfedh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+----
+<1> The name of the corresponding `Secret` object. Note that `ClientId` can alternatively refer to a `ConfigMap` object, although `clientSecret` must refer to a `Secret` object.
+<2> The OAuth 2.0 credentials contained in the specified `Secret` object.
+<3> URL used to fetch a token with the specified `clientId` and `clientSecret`.
```
```suggestion
<3> The URL used to fetch a token with the specified `clientId` and `clientSecret`.
```

bburt-rh replied:
Fixed

rolfedh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+|Authentication method|Config map field|Description
+
+|AWS Signature Version 4|`sigv4`|This method uses AWS Signature Version 4 authentication to sign requests.
+You cannot use this method simultaneously with bearer token, OAuth 2.0, or basic authentication.
```
```suggestion
You cannot use this method simultaneously with a bearer token, OAuth 2.0, or basic authentication.
```

simonpasquier replied:
```suggestion
You cannot use this method simultaneously with authorization, OAuth 2.0, or basic authentication.
```

bburt-rh replied:
fixed

rolfedh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+When you use this method, the `remoteWrite` settings in the config map must include an `authorization` `type` of `Bearer`.
+
+|OAuth 2.0|`oauth2`|An OAuth 2.0 configuration uses the client credentials grant type. 
+Prometheus fetches an access token from `tokenUrl` with the specified client ID and client secret to access the remote write endpoint using this token.
```
```suggestion
Prometheus fetches an access token from `tokenUrl` with the specified client ID and client secret to access the remote write endpoint.
```

bburt-rh replied:
deleted the final phrase per your suggestion

rolfedh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+            key: client.key <4>
+----
+<1> The name of the corresponding `Secret` object that contains the TLS authentication credentials. Note that `ca` and `cert` can alternatively refer to a `ConfigMap` object, though `keySecret` must refer to a `Secret` object.
+<2> The CA certificate for the endpoint as configured in the specified `Secret` object.
```
```suggestion
<2> The CA certificate for the endpoint, as configured in the specified `Secret` object.
```

rolfedh commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+----
+<1> The name of the corresponding `Secret` object that contains the TLS authentication credentials. Note that `ca` and `cert` can alternatively refer to a `ConfigMap` object, though `keySecret` must refer to a `Secret` object.
+<2> The CA certificate for the endpoint as configured in the specified `Secret` object.
+<3> The client certificate for the endpoint as configured in the specified `Secret` object.
```
```suggestion
<3> The client certificate for the endpoint, as configured in the specified `Secret` object.
```

bburt-rh replied:
fixed

rolfedh commented:
I have completed the [OpenShift Docs Peer Review Checklist](https://docs.google.com/document/d/12IkjTgt5uwj2etJls9mjd7LDmloW1Mn5olJabwyLNUU/edit?usp=sharing) for this PR.

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+  name: sigv4-credentials
+  namespace: openshift-monitoring
+stringData:
+  access: <AWS_access_key> <1>
+  secret: <AWS_secret_key> <2>
```
`secret` might be confused with the Kubernetes Secret resource. Maybe this?
```suggestion
  accessKey: <AWS_access_key> <1>
  secretKey: <AWS_secret_key> <2>
```

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+            key: password <3>
+----
+<1> The name of the `Secret` object that contains the authentication credentials.
+<2> The username contained in the specified `Secret` object.
+<3> The password contained in the specified `Secret` object.
```
```suggestion
<2> The key in the specified `Secret` object containing the username.
<3> The key in the specified `Secret` object containing the password.
```

bburt-rh replied:
Edited per your suggestion as follows:

```
<2> The key that contains the username  in the specified `Secret` object.
<3> The key that contains the password in the specified `Secret` object.
```

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+            key: token <2>
+----
+<1> The name of the `Secret` object that contains the authentication credentials.
+<2> The authentication token contained in the specified `Secret` object.
```
```suggestion
<2> The key in the specified `Secret` object containing the authentication token.
```

bburt-rh replied:
Edited per your suggestion as follows:

`<2> The key that contains the authentication token in the specified `Secret` object.
`

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+            param2: <parameter_2>
+----
+<1> The name of the corresponding `Secret` object. Note that `ClientId` can alternatively refer to a `ConfigMap` object, although `clientSecret` must refer to a `Secret` object.
+<2> The OAuth 2.0 credentials contained in the specified `Secret` object.
```
```suggestion
<2> The key in the specified `Secret` object containing the OAuth 2.0 credentials.
```

bburt-rh replied:
Edited per your suggestion as follows:

`<2> The key that contains the OAuth 2.0 credentials in the specified `Secret` object.
`

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+----
+<1> The AWS region.
+<2> The name of the `Secret` object containing the AWS API access credentials.
+<3> The AWS API access credentials contained in the specified `Secret` object.
```
To be consistent with the rest of the examples I would expect each field to have its own footnote.
```suggestion
<3> The key in the specified `Secret` object containing the AWS access key.
<4> The key in the specified `Secret` object containing the AWS secret key.
```

bburt-rh replied:
Added the extra calloout/footnote and edited per your suggestion as follows:

```
<3> The key that contains the AWS API access key in the specified `Secret` object.
<4> The key that contains teh AWS API secret key in the specified `Secret` object.
<5> The name of the AWS profile that is being used to authenticate.
<6> The unique identifier for the Amazon Resource Name (ARN) assigned to your role.
```

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+
+|basic authentication|`basicAuth`|Basic authentication sets the authorization header on every remote write request with the configured username and password.
+
+|bearer token authorization|`Bearer`|Bearer token authentication sets the authorization header on every remote write request using the configured token.
```
IIUC the correct entry in the configmap is `authorization` (see below in the examples).

simonpasquier replied:
```suggestion
|authorization|`authorization`|Authorization sets the Authorization header on every remote write request using the configured token.
```

bburt-rh replied:
Edited per your suggestion

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+<2> The username contained in the specified `Secret` object.
+<3> The password contained in the specified `Secret` object.
+
+.Sample YAML for bearer token authentication using a `Secret` Object
```
```suggestion
.Sample YAML for authentication with `Authorization` header using a `Secret` Object
```

bburt-rh replied:
changed to your later suggestion

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+      remoteWrite:
+      - url: "https://authorization.example.com/api/write"
+        authorization:
+          type: Bearer
```
```suggestion
          type: Bearer <x>      
 ...
 <x> Authentication type of the request (default: `Bearer`).
```

bburt-rh replied:
Added the following callout per your suggestion:

`<1> The authentication type of the request. The default value is `Bearer`.
`

simonpasquier commented on modules/monitoring-configuring-remote-write-storage.adoc:
```diff
@@ -52,89 +62,14 @@ data:
-----
+<1> The URL of the remote write endpoint.
+<2> The authentication method and credentials for the endpoint.
+Currently supported authentication methods are AWS Signature Version 4, bearer token, basic authentication, OAuth 2.0, and TLS client. 
```
```suggestion
Currently supported authentication methods are AWS Signature Version 4, authentication using HTTP `Authorization` request header, basic authentication, OAuth 2.0, and TLS client. 
```

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+|basic authentication|`basicAuth`|Basic authentication sets the authorization header on every remote write request with the configured username and password.
+
+|bearer token authorization|`Bearer`|Bearer token authentication sets the authorization header on every remote write request using the configured token.
+You can configure bearer token authentication by specifying a secret that contains the bearer token information. 
+When you use this method, the `remoteWrite` settings in the config map must include an `authorization` `type` of `Bearer`.
```
```suggestion
```

bburt-rh replied:
Deleted

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+----
+<1> The URL of the remote write endpoint.
+<2> The required configuration details for the authentication method for the endpoint.
+Currently supported authentication methods are Amazon Web Services (AWS) Signature Version 4, bearer token, basic authentication, OAuth 2.0, and TLS client. 
```
```suggestion
Currently supported authentication methods are Amazon Web Services (AWS) Signature Version 4, authorization, basic authentication, OAuth 2.0, and TLS client. 
```

bburt-rh replied:
replaced

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+= Supported remote write authentication settings
+
+You can use different methods to authenticate with a remote write endpoint.
+Currently supported authentication methods are AWS Signature Version 4, basic authentication, bearer token authorization, OAuth 2.0, and TLS client.
```
```suggestion
Currently supported authentication methods are AWS Signature Version 4, basic authentication, authorization, OAuth 2.0, and TLS client.
```

bburt-rh replied:
deleted

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+
+|OAuth 2.0|`oauth2`|An OAuth 2.0 configuration uses the client credentials grant type. 
+Prometheus fetches an access token from `tokenUrl` with the specified client ID and client secret to access the remote write endpoint using this token.
+You cannot use this method simultaneously with bearer token, AWS Signature Version 4, or basic authentication.
```
```suggestion
You cannot use this method simultaneously with authorization, AWS Signature Version 4, or basic authentication.
```

bburt-rh replied:
replaced per your suggestion

simonpasquier commented on modules/monitoring-supported-remote-write-authentication-settings.adoc:
```diff
@@ -0,0 +1,314 @@
+
+.Sample YAML for bearer token authentication using a `Secret` Object
+
+The following shows bearer token settings for a `Secret` object named `rw-bearer-auth` in the `openshift-monitoring` namespace:
```
I think that it's fine to mention bearer token from this point (to keep the example concrete).

bburt-rh replied:
@simonpasquier Would it be more accurate to say `bearer token authorization` rather than `bearer token authentication`?

simonpasquier replied:
The configuration entry is called `authorization` because this is the name of the HTTP header but this is misleading since the mechanism is really meant to authenticate the HTTP client (see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization). So while it's `authorization` in the configuration file, I believe that the documentation should use authentication/authenticate.

simonpasquier reviewed (commented):
Great! There's a bit of clarification needed between bearer token and authorization (the latter is more general).

rolfedh reviewed (approved):
Pre-approved to expedite merging.

simonpasquier reviewed (commented):
lgtm

JStickler commented:
/cherry-pick enterprise-4.11
